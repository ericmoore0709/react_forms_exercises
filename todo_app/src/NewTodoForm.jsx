import { useState } from "react";
import PropTypes from 'prop-types';

const NewTodoForm = ({ addTodo }) => {

    const INITIAL_DATA = {
        task: ''
    };

    const [formData, setFormData] = useState(INITIAL_DATA);

    const handleChange = (e) => {
        const { name, value } = e.target;

        setFormData({
            ...formData,
            [name]: value
        });
    }

    const handleSubmit = (e) => {
        e.preventDefault();
        addTodo(formData);
        setFormData(INITIAL_DATA);
    }

    return (
        <form onSubmit={handleSubmit}>
            <div className="input-container">
                <label htmlFor="task">Task</label>
                <input id="task" name="task" placeholder="Task..." value={formData.task} onChange={handleChange} />
            </div>
            <button>Add</button>
        </form>
    )
}

NewTodoForm.propTypes = {
    addTodo: PropTypes.func.isRequired
}

export default NewTodoForm;