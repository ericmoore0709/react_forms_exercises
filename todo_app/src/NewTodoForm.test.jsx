import { expect, it, vi } from 'vitest';
import { render, fireEvent, screen } from '@testing-library/react';
import NewTodoForm from './NewTodoForm';

it('should not smoke', () => {
    render(<NewTodoForm addTodo={vi.fn()} />);
});

it('should match snapshot', () => {
    const { asFragment } = render(<NewTodoForm addTodo={vi.fn()} />);
    expect(asFragment()).toMatchSnapshot();
});

it('should fire addTodo on submit', () => {
    const addMock = vi.fn();
    render(<NewTodoForm addTodo={addMock} />);
    fireEvent.click(screen.getByRole('button'));
    expect(addMock).toBeCalled();
});

it('should change task when input is changed', () => {
    render(<NewTodoForm addTodo={vi.fn()} />);
    const taskInput = screen.getByLabelText('Task');

    // change task value
    fireEvent.change(taskInput, {
        target: { value: "Test task" },
    });
    expect(taskInput.value).toBe('Test task');
});

it('should clear form data on submit', () => {
    const addMock = vi.fn();
    const { container } = render(<NewTodoForm addTodo={addMock} />);

    const taskInput = screen.getByLabelText('Task');

    // change task
    fireEvent.change(taskInput, {
        target: { value: "New task" },
    });

    fireEvent.click(container.querySelector('button'));
    expect(taskInput.value).toBe('');
});
