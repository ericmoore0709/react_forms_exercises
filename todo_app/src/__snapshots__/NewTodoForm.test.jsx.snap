// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <form>
    <div
      class="input-container"
    >
      <label
        for="task"
      >
        Task
      </label>
      <input
        id="task"
        name="task"
        placeholder="Task..."
        value=""
      />
    </div>
    <button>
      Add
    </button>
  </form>
</DocumentFragment>
`;
